lib/spack/spack/build_environment.py

```diff
@@ -452,6 +451,17 @@ def set_build_environment_variables(pkg, env, dirty):
     return env
 
 
+def determine_number_of_jobs(
+        cap,
+        command_line=spack.config.get('config:build_jobs', scope='command_line'),
+        default=spack.config.get('config:build_jobs')):
+    return command_line if command_line is not None else min(default, cap)
```

Comment:
config accesses in the defaults of a method signature will break Spack in very difficult ways to debug. 
```suggestion
def determine_number_of_jobs(
        cap,
        command_line=None,
        default=None):
    command_line = command_line or spack.config.get('config:build_jobs', scope='command_line')
    default = default or spack.config.get('config:build_jobs')
    return command_line if command_line is not None else min(default, cap)
```

Also, this method is somewhat dense. I was able to wade through it, but a comment clarifyiing how the ternary works will make it easier for future maintainers.

Reply:
I've tried to simplify it a bit, and added a docstring